frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx

```diff
@@ -0,0 +1,97 @@
+import * as React from 'react';
+import { MenuToggle } from '@patternfly/react-core';
+import { EllipsisVIcon } from '@patternfly/react-icons';
+import * as _ from 'lodash';
+import { useTranslation } from 'react-i18next';
+import { Action } from '@console/dynamic-plugin-sdk';
+import { checkAccess } from '@console/internal/components/utils';
+import ActionServiceProvider from './ActionServiceProvider';
+import ActionMenuRenderer from './menu/ActionMenuRenderer';
+import { ActionContext, ActionMenuVariant } from './types';
+
+type LazyActionMenuProps = {
+  context: ActionContext;
+  variant?: ActionMenuVariant;
+  label?: string;
+  isDisabled?: boolean;
+};
+
+const LazyActionMenu: React.FC<LazyActionMenuProps> = ({
+  context,
+  variant = ActionMenuVariant.KEBAB,
+  label,
+  isDisabled,
+}) => {
+  const { t } = useTranslation();
+  const isKebabVariant = variant === ActionMenuVariant.KEBAB;
+  const [active, setActive] = React.useState<boolean>(false);
+  const [initActionLoader, setInitActionLoader] = React.useState<boolean>(false);
+  const toggleRef = React.useRef<HTMLButtonElement>();
+  const toggleRefCb = React.useCallback(() => toggleRef.current, []);
+  const toggleLabel = label || t('console-shared~Actions');
+
+  const toggleMenu = () => setActive((value) => !value);
+
+  const hideMenu = () => {
+    toggleRef.current?.focus();
+    setActive(false);
+  };
+
+  const handleRequestClose = (e?: MouseEvent) => {
+    if (!e || !toggleRef.current?.contains(e.target as Node)) {
+      hideMenu();
+    }
+  };
+
+  const handleHover = React.useCallback(() => {
+    setInitActionLoader(true);
+  }, []);
+
+  const runAccessChecks = (actions: Action[]) => {
+    // Check access after loading actions from service over a kebab to minimize flicker when opened.
+    // This depends on `checkAccess` being memoized.
+    _.each(actions, (action: Action) => {
+      if (action.accessReview) {
+        checkAccess(action.accessReview);
+      }
+    });
+  };
+
+  return (
+    <div>
+      <MenuToggle
+        variant={variant}
+        innerRef={toggleRef}
+        isExpanded={active}
+        isDisabled={isDisabled}
+        aria-expanded={active}
+        aria-label={toggleLabel}
+        aria-haspopup="true"
+        data-test-id="menu-toggle-button"
+        onClick={toggleMenu}
+        {...(isKebabVariant ? { onFocus: handleHover, onMouseEnter: handleHover } : {})}
+      >
+        {isKebabVariant ? <EllipsisVIcon /> : toggleLabel}
+      </MenuToggle>
+      {initActionLoader && (
+        <ActionServiceProvider context={context}>
+          {({ actions, options, loaded }) => {
+            if (!loaded) return null;
+            runAccessChecks(actions);
```

Comment:
This should be in a useEffect. Maybe create a separate component to manage this.

Reply:
Done.